File: scalechain-chain/src/test/scala/io/scalechain/blockchain/chain/TransactionMagnetSpec.scala
package io.scalechain.blockchain.chain

import java.io.File

import io.scalechain.blockchain.script.HashSupported
import io.scalechain.blockchain.transaction.TransactionTestDataTrait
import org.scalatest._
import HashSupported._

/**
  * Created by kangmo on 6/16/16.
  */
class TransactionMagnetSpec extends BlockchainTestTrait with TransactionTestDataTrait with ShouldMatchers {

  this: Suite =>

  val testPath = new File("./target/unittests-TransactionMagnetSpec/")

  import BlockSampleData._
  import BlockSampleData.Tx._
  import BlockSampleData.Block._

  var tm : TransactionMagnet = null

  override def beforeEach() {
    // initialize a test.

    super.beforeEach()

    // put the genesis block
    chain.putBlock(env.GenesisBlockHash, env.GenesisBlock)

    tm = chain.txMagnet
  }

  override def afterEach() {
    super.afterEach()

    // finalize a test.
    tm = null
  }

  "markOutputSpent" should "" in {
  }

  "markOutputUnspent" should "" in {
  }

  "markAllOutputsUnspent" should "" in {
  }

  "detachTransactionInput" should "" in {
  }

  "detachTransactionInputs" should "" in {
  }

  "detachTransaction" should "" in {
  }

  "attachTransactionInput" should "" in {
  }

  "attachTransactionInputs" should "" in {
  }

  "attachTransaction" should "" in {
  }
}

File: scalechain-chain/src/test/kotlin/io/scalechain/blockchain/chain/TransactionMagnetSpec.kt
package io.scalechain.blockchain.chain

import io.kotlintest.KTestJUnitRunner
import io.kotlintest.matchers.Matchers
import io.scalechain.blockchain.script.hash
import java.io.File

import io.scalechain.blockchain.transaction.TransactionTestInterface
import org.junit.runner.RunWith

/**
  * Created by kangmo on 6/16/16.
  */
@RunWith(KTestJUnitRunner::class)
class TransactionMagnetSpec : BlockchainTestTrait(), TransactionTestInterface, Matchers {
  override val testPath = File("./target/unittests-TransactionMagnetSpec/")

  lateinit var tm : TransactionMagnet

  override fun beforeEach() {
    // initialize a test.

    super.beforeEach()

    // put the genesis block
    chain.putBlock(db, env().GenesisBlockHash, env().GenesisBlock)

    tm = chain.txMagnet
  }

  override fun afterEach() {
    super.afterEach()

  }

  init {

    "markOutputSpent" should "" {
/*
      val data = TransactionSampleData(db)
      val B = data.Block
      val T = data.Tx
*/
    }

    "markOutputUnspent" should "" {
    }

    "markAllOutputsUnspent" should "" {
    }

    "detachTransactionInput" should "" {
    }

    "detachTransactionInputs" should "" {
    }

    "detachTransaction" should "" {
    }

    "attachTransactionInput" should "" {
    }

    "attachTransactionInputs" should "" {
    }

    "attachTransaction" should "attach transactions in order" {
      val data = TransactionSampleData(db)
      val B = data.Block
      val T = data.Tx

      chain.putBlock(db, data.env().GenesisBlockHash, data.env().GenesisBlock)
      chain.putBlock(db, B.BLK01.header.hash(), B.BLK01)
      chain.putBlock(db, B.BLK02.header.hash(), B.BLK02)
      chain.putBlock(db, B.BLK03.header.hash(), B.BLK03)

      listOf(T.TX04_01, T.TX04_02, T.TX04_03, T.TX04_04).forEach { tx ->
        //println("hash : ${tx.transaction.hash()}")
        tm.attachTransaction(db, tx.transaction.hash(), tx.transaction, false)
      }
    }
  }
}
